### PR TITLE
Addition of multiclass adaboost algorithm

### PR DESCRIPTION
**Pull request by orazve**
Added multiclass adaboost: added resultsToEvaluate field to classifier::Parameter, all classifiers are moved to the next interface, logistic regression resultsToCompute parameter is deprecated, removed logistic_regression::prediction::Result class.

### CODE REVIEW

File: algorithms/kernel/kernel.h [hunk 1]
```diff
@@ -54,13 +54,13 @@
         ((KernelClass<templateArguments, cpu> *)(_kernel))->method(__VA_ARGS__);
 
 #define __DAAL_INSTANTIATE_DISPATCH_IMPL(ContainerTemplate, Mode, ClassName, BaseClassName, ...) \
-extern template class DAAL_KERNEL_SSE2_CONTAINER1(ContainerTemplate, __VA_ARGS__);               \
-extern template class DAAL_KERNEL_SSSE3_CONTAINER1(ContainerTemplate, __VA_ARGS__);              \
-extern template class DAAL_KERNEL_SSE42_CONTAINER1(ContainerTemplate, __VA_ARGS__);              \
-extern template class DAAL_KERNEL_AVX_CONTAINER1(ContainerTemplate, __VA_ARGS__);                \
-extern template class DAAL_KERNEL_AVX2_CONTAINER1(ContainerTemplate, __VA_ARGS__);               \
-extern template class DAAL_KERNEL_AVX512_MIC_CONTAINER1(ContainerTemplate, __VA_ARGS__);         \
-extern template class DAAL_KERNEL_AVX512_CONTAINER1(ContainerTemplate, __VA_ARGS__);             \
```
Comment (SmirnovEgorRu) [inline algorithms/kernel/kernel.h]: What is the reason for the changes?

File: lang_service/java/com/intel/daal/algorithms/regression/predict_result.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ * Class:     com_intel_daal_algorithms_regression_prediction_PredictionResult
+ * Method:    cGetResult
+ * Signature: (JI)J
+ */
+JNIEXPORT jlong JNICALL Java_com_intel_daal_algorithms_regression_prediction_PredictionResult_cGetResult
+(JNIEnv *env, jobject thisObj, jlong algAddr )
+{
+    SerializationIfacePtr *ptr = new SerializationIfacePtr();
+
+    //    if(cmode == jBatch)
```
Comment (prutskov) [inline lang_service/java/com/intel/daal/algorithms/regression/predict_result.cpp]: delete comments

File: algorithms/kernel/kernel.h (hunk 1 above)
Comment (orazve) [reply algorithms/kernel/kernel.h]: Together with changes in include/services/daal_kernel_defines.h, it fixes a bug of building on one particular processor core. I can separate this change from my pull request.

File: examples/cpp/DAALExamples.sln [hunk 3]
```diff
@@ -396,6 +396,20 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "dbscan_dense_batch", "vcpro
 EndProject
 Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "dbscan_dense_distr", "vcproj\dbscan_dense_distr\dbscan_dense_distr.vcxproj", "{F87748FC-F17D-4CDE-B697-69DFDECFADF6}"
 EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "adaboost_samme_multi_class_batch", "vcproj\adaboost_samme_multi_class_batch\adaboost_samme_multi_class_batch.vcxproj", "{8E460210-47C5-4046-B4F3-69A3691D11CD}"
```
[the remark this answers is not in the document]
Comment (orazve) [reply examples/cpp/DAALExamples.sln]: It was already run, no changes needed.

File: examples/java/com/intel/daal/examples/boosting/brownboost/BrownBoostDenseBatch.java [hunk 4]
```diff
@@ -22,7 +22,7 @@
  */
 
 /**
- * <a name="DAAL-EXAMPLE-JAVA-BROWNBOOSTBATCH">
+ * <a name="DAAL-EXAMPLE-JAVA-BROWNBOOST_DENSE_BATCH">
```
[the remark this answers is not in the document]
Comment (orazve) [reply examples/java/com/intel/daal/examples/boosting/brownboost/BrownBoostDenseBatch.java]: Mistakenly changed, fixed

File: examples/java/com/intel/daal/examples/boosting/logitboost/LogitBoostDenseBatch.java [hunk 5]
```diff
@@ -25,7 +25,7 @@
  */
 
 /**
- * <a name="DAAL-EXAMPLE-JAVA-LOGITBOOSTBATCH">
+ * <a name="DAAL-EXAMPLE-JAVA-LOGITBOOST_DENSE_BATCH">
```
[the remark this answers is not in the document]
Comment (orazve) [reply examples/java/com/intel/daal/examples/boosting/logitboost/LogitBoostDenseBatch.java]: Mistakenly changed, fixed

File: lang_service/java/com/intel/daal/algorithms/regression/predict_result.cpp (hunk 2 above)
Comment (orazve) [reply lang_service/java/com/intel/daal/algorithms/regression/predict_result.cpp]: done

File: lang_service/java/com/intel/daal/algorithms/classifier/predict_result.cpp [hunk 6]
```diff
@@ -51,10 +50,13 @@ JNIEXPORT jlong JNICALL Java_com_intel_daal_algorithms_classifier_prediction_Pre
 {
     SerializationIfacePtr *ptr = new SerializationIfacePtr();
 
+    //    if(cmode == jBatch)
```
[the remark this answers is not in the document]
Comment (orazve) [reply lang_service/java/com/intel/daal/algorithms/classifier/predict_result.cpp]: done

File: lang_interface/java/com/intel/daal/algorithms/adaboost/Parameter.java [hunk 7]
```diff
@@ -69,12 +86,32 @@ public long getMaxIterations() {
         return cGetMaxIterations(this.cObject);
     }
 
-    private native void cSetAccuracyThreshold(long parAddr, double acc);
+    /**
+     * Sets the accuracy of the AdaBoost training algorithm
```
[the remark this answers is not in the document]
Comment (orazve) [reply lang_interface/java/com/intel/daal/algorithms/adaboost/Parameter.java]: they are different methods, fixed comment

File: lang_interface/java/com/intel/daal/algorithms/adaboost/Model.java [hunk 8]
```diff
@@ -39,16 +39,61 @@ public Model(DaalContext context, long cModel) {
...
+     */
+    public long getNumberOfFeatures() {
+        return cGetNumberOfFeatures(this.cObject);
+    }
+
     /**
      * Returns the numeric table that contains the array of weights of weak learners constructed
-     * during training of the AdaBoost algorithm.
-     * The size of the array equals the number of weak learners
+     * during training of the BrownBoost algorithm. The size of the array equals the number of weak learners
```
[the remark this answers is not in the document]
Comment (orazve) [reply lang_interface/java/com/intel/daal/algorithms/adaboost/Model.java]: fixed

File: lang_interface/java/com/intel/daal/algorithms/stump/classification/training/TrainingResult.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+import com.intel.daal.utils.*;
+import com.intel.daal.algorithms.ComputeMode;
+import com.intel.daal.algorithms.stump.classification.Model;
+import com.intel.daal.algorithms.classifier.training.TrainingResultId;
+import com.intel.daal.services.DaalContext;
+
+/**
+ * <a name="DAAL-CLASS-ALGORITHMS__STUMP__CLASSIFICATION__TRAINING__TRAININGRESULT"></a>
+ * @brief Provides methods to access final results obtained with the compute() method of the AdaBoost training algorithm in the batch processing mode
```
[the remark this answers is not in the document]
Comment (orazve) [reply lang_interface/java/com/intel/daal/algorithms/stump/classification/training/TrainingResult.java]: fixed

**Comment by averbukh** [comment]
make test

**Comment by orazve** [comment]
> "make  test"
won

File: algorithms/kernel/adaboost/adaboost_predict_impl.i [hunk 10]
```diff
@@ -37,10 +40,209 @@ namespace prediction
...
+    learnerInput->set(classifier::prediction::data, xTable);
+
+    classifier::prediction::ResultPtr predictionRes(new classifier::prediction::Result());
+    predictionRes->set(classifier::prediction::prediction, rWeakTable);
+    DAAL_CHECK_STATUS(s, learnerPredict->setResult(predictionRes));
+
+    const algorithmFPType zero = (algorithmFPType)0.0;
+
+    /* Initialize array of prediction results */
+    for (size_t j = 0; j < nVectors; j++)
```
Comment (ShvetsKS) [inline algorithms/kernel/adaboost/adaboost_predict_impl.i]: can we use 'service_memset' here?

File: algorithms/kernel/adaboost/adaboost_predict_impl.i [hunk 11]
```diff
@@ -37,10 +40,209 @@ namespace prediction
...
-services::Status AdaBoostPredictKernel<method, algorithmFPType, cpu>::compute(const NumericTablePtr& xTable,
-    const Model *m, const NumericTablePtr& rTable, const Parameter *par)
+services::Status AdaBoostPredictKernelNew<method, algorithmFPType, cpu>::computeSammeProbability(
+    const algorithmFPType *p, const size_t nVectors, const size_t nClasses, algorithmFPType *h)
+{
+    algorithmFPType *pLog = h;
+    TArray<algorithmFPType, cpu> pSumLogArray(nVectors);
+    DAAL_CHECK(pSumLogArray.get(), services::ErrorMemoryAllocationFailed);
+    algorithmFPType *pSumLog = pSumLogArray.get();
+    for(size_t i = 0; i < nVectors; i++)
```
Comment (ShvetsKS) [inline algorithms/kernel/adaboost/adaboost_predict_impl.i]: use 'service_memset' if possible

Comment (ShvetsKS) [reply algorithms/kernel/adaboost/adaboost_predict_impl.i]: the same comment is applicable for similar initializations.

File: algorithms/kernel/adaboost/adaboost_model_fpt.cpp [hunk 12]
```diff
@@ -29,23 +29,28 @@ namespace algorithms
...
 
 template <typename modelFPType>
-DAAL_EXPORT Model::Model(size_t nFeatures, modelFPType dummy, services::Status &st) : boosting::Model(nFeatures, st)
+DAAL_EXPORT Model::Model(size_t nFeatures, modelFPType dummy, services::Status &st) :
+    _nFeatures(nFeatures),
+    _models(new data_management::DataCollection())
 {
-    if (!st) { return; }
+    if (!_models) { st.add(services::ErrorMemoryAllocationFailed); }
     _alpha = data_management::HomogenNumericTable<modelFPType>::create(NULL, 1, 0, &st);
```
Comment (averbukh) [inline algorithms/kernel/adaboost/adaboost_model_fpt.cpp]: Could it be moved into initialization?

File: algorithms/kernel/classifier/classifier_predict_fpt.cpp [hunk 13]
```diff
@@ -45,13 +48,47 @@ template <typename algorithmFPType>
...
+{
+    services::Status st;
+    const Parameter *par = static_cast<const Parameter *>(parameter);
+    DAAL_CHECK(par, services::ErrorNullParameterNotSupported);
+
+    const size_t nRows = (static_cast<const InputIface *>(input))->getNumberOfRows();
+    const size_t nClasses = par->nClasses;
+
+    if(par->resultsToEvaluate & computeClassesLabels)
+        set(prediction, HomogenNumericTable<algorithmFPType>::create(1, nRows, NumericTableIface::doAllocate, &st));
```
Comment (ShvetsKS) [inline algorithms/kernel/classifier/classifier_predict_fpt.cpp]: Why don`t you use create method with default value as bellow?

File: algorithms/kernel/decision_tree/decision_tree_classification_predict_dense_default_batch_impl.i [hunk 14]
```diff
@@ -53,12 +53,11 @@ using namespace decision_tree::internal;
 
 template<typename algorithmFPType, CpuType cpu>
 services::Status DecisionTreePredictKernel<algorithmFPType, defaultDense, cpu>::
-    compute(const NumericTable * x, const classifier::Model * m, NumericTable * y, const daal::algorithms::Parameter * par)
+    compute(const NumericTable * x, const classifier::Model * m, NumericTable * y, NumericTable * p, const daal::algorithms::Parameter * par)
 {
     typedef daal::services::internal::SignBit<algorithmFPType, cpu> SignBitType;
 
     DAAL_ASSERT(x);
-    DAAL_ASSERT(y);
```
Comment (ShvetsKS) [inline algorithms/kernel/decision_tree/decision_tree_classification_predict_dense_default_batch_impl.i]: Why it was deleted?

File: include/algorithms/logistic_regression/logistic_regression_predict_types.h [hunk 15]
```diff
@@ -218,18 +218,14 @@ class DAAL_EXPORT Result : public algorithms::classifier::prediction::Result
...
+        return daal::algorithms::classifier::prediction::interface1::Result::serialImpl<Archive, onDeserialize>(arch);
     }
 };
 typedef services::SharedPtr<Result> ResultPtr;
 typedef services::SharedPtr<const Result> ResultConstPtr;
 
 } // namespace interface1
-using interface1::Parameter;
 using interface1::Input;
-using interface1::Result;
```
Comment (ShvetsKS) [inline include/algorithms/logistic_regression/logistic_regression_predict_types.h]: Static backward compatibility will be broken due to this change. I thought interface2 should be added for logistic regression Result to reduce changes related to static backward compatibility.

File: include/algorithms/logistic_regression/logistic_regression_predict.h [hunk 16]
```diff
@@ -193,8 +193,156 @@ class DAAL_EXPORT Batch : public classifier::prediction::Batch
...
+ */
+template<typename algorithmFPType = DAAL_ALGORITHM_FP_TYPE, Method method = defaultDense>
+class DAAL_EXPORT Batch : public classifier::prediction::interface2::Batch
+{
+public:
+    typedef classifier::prediction::interface2::Batch super;
+
+    typedef algorithms::logistic_regression::prediction::Input     InputType;
+    typedef algorithms::classifier::interface2::Parameter          ParameterType;
+    typedef algorithms::classifier::prediction::interface2::Result ResultType;
```
Comment (ShvetsKS) [inline include/algorithms/logistic_regression/logistic_regression_predict.h]: Static backward compatibility will be broken due to this change. I thought interface2 should be added for logistic regression Result to reduce changes related to static backward compatibility.

File: algorithms/kernel/kernel.h (hunk 1 above)
Comment (ShvetsKS) [reply algorithms/kernel/kernel.h]: It was modified by Andrey Gilev to reduce the size of the library with avoiding unnecessary template instantiation.  Need to discuss this solution with Gilev Andrey.

File: algorithms/kernel/logistic_regression/logistic_regression_predict_types.cpp [hunk 17]
```diff
@@ -93,49 +91,11 @@ services::Status Input::check(const daal::algorithms::Parameter *parameter, int
     const daal::algorithms::logistic_regression::internal::ModelImpl* pModel =
         static_cast<const daal::algorithms::logistic_regression::internal::ModelImpl*>(m.get());
     DAAL_ASSERT(pModel);
-    const daal::algorithms::logistic_regression::prediction::Parameter* pPrm = static_cast<const daal::algorithms::logistic_regression::prediction::Parameter*>(parameter);
+    const daal::algorithms::logistic_regression::prediction::interface1::Parameter* pPrm = static_cast<const daal::algorithms::logistic_regression::prediction::interface1::Parameter*>(parameter);
```
Comment (ShvetsKS) [inline algorithms/kernel/logistic_regression/logistic_regression_predict_types.cpp]: Dependency on deprecated 'interface1::Parameter' inside still valid interface1::Input which is used in interface2::Batch.

File: algorithms/kernel/logistic_regression/logistic_regression_predict_types.cpp [hunk 18]
```diff
@@ -93,49 +91,11 @@ services::Status Input::check(const daal::algorithms::Parameter *parameter, int
...
     const size_t nBetaPerClass = get(classifier::prediction::data)->getNumberOfColumns() + 1;
     return checkNumericTable(pModel->getBeta().get(), betaStr(), 0, 0, nBetaPerClass, pPrm->nClasses == 2 ? 1 : pPrm->nClasses);
 }
 
-Result::Result() : algorithms::classifier::prediction::Result(lastResultNumericTableId + 1) {}
```
Comment (ShvetsKS) [inline algorithms/kernel/logistic_regression/logistic_regression_predict_types.cpp]: See comment above regarding Result class depreciation

File: include/algorithms/boosting/boosting_model.h [hunk 19]
```diff
@@ -57,7 +57,7 @@ namespace interface1
  * \snippet boosting/boosting_model.h Parameter source code
  */
 /* [Parameter source code] */
-struct DAAL_EXPORT Parameter : public classifier::Parameter
+struct DAAL_EXPORT Parameter : public classifier::interface1::Parameter
```
Comment (ShvetsKS) [inline include/algorithms/boosting/boosting_model.h]: Does it mean that this class is deprecated as there is no interface2::Parameter?

File: include/algorithms/boosting/boosting_predict.h [hunk 20]
```diff
@@ -74,10 +74,10 @@ namespace interface1
  *      - \ref classifier::prediction::interface1::Input "classifier::prediction::Input" class
  *      - \ref classifier::prediction::interface1::Result "classifier::prediction::Result" class
  */
-class Batch : public classifier::prediction::Batch
+class Batch : public classifier::prediction::interface1::Batch
```
Comment (ShvetsKS) [inline include/algorithms/boosting/boosting_predict.h]: Does it mean that this class is deprecated as there is no interface2::Batch?

File: include/algorithms/boosting/boosting_training_batch.h [hunk 21]
```diff
@@ -68,10 +68,10 @@ namespace interface1
  *      - \ref classifier::training::interface1::Input "classifier::training::Input" class
  *      - \ref classifier::training::interface1::Result "classifier::training::Result" class
  */
-class DAAL_EXPORT Batch : public classifier::training::Batch
+class DAAL_EXPORT Batch : public classifier::training::interface1::Batch
```
Comment (ShvetsKS) [inline include/algorithms/boosting/boosting_training_batch.h]: Does it mean that this class is deprecated as there is no interface2::Batch?

File: include/algorithms/logistic_regression/logistic_regression_predict_types.h [hunk 22]
```diff
@@ -61,9 +61,9 @@ enum Method
 */
 enum ResultToComputeId
 {
-    computeClassesLabels = 0x00000001ULL,
-    computeClassesProbabilities = 0x00000002ULL,
-    computeClassesLogProbabilities = 0x00000004ULL
+    computeClassesLabels = classifier::computeClassesLabels,
+    computeClassesProbabilities = classifier::computeClassesProbabilities,
```
Comment (ShvetsKS) [inline include/algorithms/logistic_regression/logistic_regression_predict_types.h]: Sorry for too delay.
seems we have to change computeClassesLabels to computeClassLabels according to Andrey Nikolaev comment regarding the naming.
for probabilities and log probabilities it`s applicable also.

File: include/algorithms/weak_learner/weak_learner_model.h [hunk 23]
```diff
@@ -51,7 +51,7 @@ namespace interface1
  * \snippet weak_learner/weak_learner_model.h Parameter source code
  */
 /* [Parameter source code] */
-class Parameter : public classifier::Parameter
+class Parameter : public classifier::interface1::Parameter
```
Comment (ShvetsKS) [inline include/algorithms/weak_learner/weak_learner_model.h]: Does it mean that Parameter class is deprecated now as there is no interface2::Parameter?

File: include/algorithms/weak_learner/weak_learner_predict.h [hunk 24]
```diff
@@ -69,10 +69,10 @@ namespace interface1
  *      - \ref interface1::Input "Input" class
  *      - \ref interface1::Result "Result" class
  */
-class Batch : public classifier::prediction::Batch
+class Batch : public classifier::prediction::interface1::Batch
```
Comment (ShvetsKS) [inline include/algorithms/weak_learner/weak_learner_predict.h]: Does it mean that Batch class is deprecated now as there is no interface2::Batch?

File: include/algorithms/weak_learner/weak_learner_training_batch.h [hunk 25]
```diff
@@ -57,10 +57,10 @@ namespace interface1
  *      - \ref interface1::Model "Model" class
  *      - Result class
  */
-class DAAL_EXPORT Batch : public classifier::training::Batch
+class DAAL_EXPORT Batch : public classifier::training::interface1::Batch
```
Comment (ShvetsKS) [inline include/algorithms/weak_learner/weak_learner_training_batch.h]: Does it mean that Batch class is deprecated now as there is no interface2::Batch?

File: include/algorithms/weak_learner/weak_learner_training_types.h [hunk 26]
```diff
@@ -59,7 +59,7 @@ namespace interface1
  * \brief Provides methods to access final results obtained with compute() method of Batch
  *        or finalizeCompute() method of Online and Distributed weak learners algorithms
  */
-class DAAL_EXPORT Result : public daal::algorithms::classifier::training::Result
+class DAAL_EXPORT Result : public daal::algorithms::classifier::training::interface1::Result
```
Comment (ShvetsKS) [inline include/algorithms/weak_learner/weak_learner_training_types.h]: Does it mean that Result class is deprecated now as there is no interface2::Result?

File: include/algorithms/boosting/boosting_model.h (hunk 19 above)
Comment (orazve) [reply include/algorithms/boosting/boosting_model.h]: yes, full boosting:: namespace is deprecated

File: include/algorithms/decision_tree/decision_tree_regression_training_types.h [hunk 27]
```diff
@@ -75,6 +75,7 @@ enum InputId
 {
     data                            = algorithms::regression::training::data,               /*!< %Input data table */
     dependentVariables              = algorithms::regression::training::dependentVariables, /*!< Values of the dependent variable for the input data */
+    weights                         = algorithms::regression::training::weights,            /*!< Optional. Weights of the observations in the training data set */
```
Comment (ShvetsKS) [inline include/algorithms/decision_tree/decision_tree_regression_training_types.h]: This change possible broke dynamic backward compatibility.

File: algorithms/kernel/adaboost/adaboost_model.cpp [hunk 28]
```diff
@@ -22,6 +22,8 @@
 */
 
 #include "algorithms/boosting/adaboost_model.h"
+#include "algorithms/stump/stump_classification_training_batch.h"
```
Comment (orazve) [inline algorithms/kernel/adaboost/adaboost_model.cpp]: comment for reopening pull request

File: algorithms/kernel/adaboost/adaboost_model_fpt.cpp (hunk 12 above)
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_model_fpt.cpp]: no, there is no initialization function for adaboost model

File: algorithms/kernel/decision_tree/decision_tree_classification_predict_dense_default_batch_impl.i (hunk 14 above)
Comment (orazve) [reply algorithms/kernel/decision_tree/decision_tree_classification_predict_dense_default_batch_impl.i]: because now it can be null, if e.g. a user set only probabilities to compute

File: algorithms/kernel/classifier/classifier_predict_fpt.cpp (hunk 13 above)
Comment (orazve) [reply algorithms/kernel/classifier/classifier_predict_fpt.cpp]: will fix

File: algorithms/kernel/logistic_regression/logistic_regression_predict_types.cpp (hunk 17 above)
Comment (orazve) [reply algorithms/kernel/logistic_regression/logistic_regression_predict_types.cpp]: Fixed

File: include/algorithms/boosting/boosting_predict.h (hunk 20 above)
Comment (orazve) [reply include/algorithms/boosting/boosting_predict.h]: boosting namespace is fully deprecated

File: include/algorithms/boosting/boosting_training_batch.h (hunk 21 above)
Comment (orazve) [reply include/algorithms/boosting/boosting_training_batch.h]: boosting namespace is fully deprecated

File: include/algorithms/logistic_regression/logistic_regression_predict_types.h (hunk 22 above)
Comment (orazve) [reply include/algorithms/logistic_regression/logistic_regression_predict_types.h]: fixed

File: include/algorithms/weak_learner/weak_learner_model.h (hunk 23 above)
Comment (orazve) [reply include/algorithms/weak_learner/weak_learner_model.h]: weak_learner namespace is fully deprecated

File: include/algorithms/weak_learner/weak_learner_predict.h (hunk 24 above)
Comment (orazve) [reply include/algorithms/weak_learner/weak_learner_predict.h]: weak_learner namespace is fully deprecated

File: include/algorithms/weak_learner/weak_learner_training_batch.h (hunk 25 above)
Comment (orazve) [reply include/algorithms/weak_learner/weak_learner_training_batch.h]: weak_learner namespace is fully deprecated

File: include/algorithms/weak_learner/weak_learner_training_types.h (hunk 26 above)
Comment (orazve) [reply include/algorithms/weak_learner/weak_learner_training_types.h]: weak_learner namespace is fully deprecated

File: include/algorithms/decision_tree/decision_tree_regression_training_types.h (hunk 27 above)
Comment (orazve) [reply include/algorithms/decision_tree/decision_tree_regression_training_types.h]: fixed

File: algorithms/kernel/adaboost/adaboost_predict_batch_container.h [hunk 29]
```diff
@@ -36,10 +36,12 @@ namespace adaboost
 {
 namespace prediction
 {
+namespace interface2
+{
 template<typename algorithmFPType, Method method, CpuType cpu>
 BatchContainer<algorithmFPType, method, cpu>::BatchContainer(daal::services::Environment::env *daalEnv)
 {
-    __DAAL_INITIALIZE_KERNELS(internal::AdaBoostPredictKernel, method, algorithmFPType);
+    __DAAL_INITIALIZE_KERNELS(internal::AdaBoostPredictKernelNew, method, algorithmFPType);
```
[the remark this answers is not in the document]
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_predict_batch_container.h]:
```
AdaBoostPredictKernel->I1AdaBoostPredictKernel
AdaBoostPredictKernelNew->AdaBoostPredictKernel
```

File: algorithms/kernel/adaboost/adaboost_predict_impl.i [hunk 30]
```diff
@@ -37,10 +40,209 @@ namespace prediction
 namespace internal
 {
 using namespace daal::internal;
+template <Method method, typename algorithmFPType, CpuType cpu>
+services::Status AdaBoostPredictKernelNew<method, algorithmFPType, cpu>::computeImpl(const NumericTablePtr &xTable,
+        const Model *m, size_t nWeakLearners, const algorithmFPType *alpha, algorithmFPType *r, const Parameter *par)
+{
+    const size_t nVectors  = xTable->getNumberOfRows();
+    Model *boostModel = const_cast<Model *>(m);
+    Parameter *parameter = const_cast<Parameter *>(par);
```
[the remark this answers is not in the document]
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_predict_impl.i]: ok

File: algorithms/kernel/adaboost/adaboost_predict_impl.i (hunk 10 above)
[the remark this answers is not in the document]
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_predict_impl.i]: I used here service_memset. It doesn't add parallel code, but it uses the common instrument.

File: algorithms/kernel/adaboost/adaboost_predict_impl.i [hunk 31]
```diff
@@ -37,10 +40,209 @@ namespace prediction
...
+    TArray<algorithmFPType, cpu> pSumLogArray(nVectors);
+    DAAL_CHECK(pSumLogArray.get(), services::ErrorMemoryAllocationFailed);
+    algorithmFPType *pSumLog = pSumLogArray.get();
+    for(size_t i = 0; i < nVectors; i++)
+    {
+        pSumLog[i] = 0;
+    }
+
+    algorithmFPType eps = services::internal::EpsilonVal<algorithmFPType>::get();
+    for(size_t i = 0; i < nVectors * nClasses; i++)
```
[the remark this answers is not in the document]
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_predict_impl.i]: It could be any

File: algorithms/kernel/adaboost/adaboost_predict_impl.i [hunk 32]
```diff
@@ -37,10 +40,209 @@ namespace prediction
...
+        }
+    }
+    for(size_t k = 0; k < nClasses; k++)
+    {
+        for(size_t m = 0; m < nWeakLearners; m++)
+        {
+            const algorithmFPType *rWeak = weakPredictions[m]->getArray();
+            for (size_t i = 0; i < nVectors; i++)
+            {
+                if(method == samme)
```
[the remark this answers is not in the document]
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_predict_impl.i]: why?

File: algorithms/kernel/brownboost/brownboost_predict_impl.i [hunk 33]
```diff
@@ -41,6 +41,92 @@ using namespace daal::internal;
...
+        ReadColumns<algorithmFPType, cpu> mtAlpha(*boostModel->getAlpha(), 0, 0, nWeakLearners);
+        DAAL_CHECK_BLOCK_STATUS(mtAlpha);
+        DAAL_ASSERT(mtAlpha.get());
+        DAAL_CHECK_STATUS(s, this->compute(xTable, m, nWeakLearners, mtAlpha.get(), r, par));
+    }
+
+    brownboost::interface1::Parameter *parameter = const_cast<brownboost::interface1::Parameter *>(par);
+    const algorithmFPType error = parameter->accuracyThreshold;
+    const algorithmFPType zero = (algorithmFPType)0.0;
+    if(error != zero)
```
[the remark this answers is not in the document]
Comment (orazve) [reply algorithms/kernel/brownboost/brownboost_predict_impl.i]: it is a constant which set by user, there is no warning comparison zero and "almost zero"

File: algorithms/kernel/decision_tree/decision_tree_classification_training_input.cpp [hunk 34]
```diff
@@ -63,9 +63,18 @@ services::Status Input::check(const daal::algorithms::Parameter * parameter, int
 services::Status Input::checkImpl(const daal::algorithms::Parameter * parameter) const
 {
     services::Status s;
-    const decision_tree::classification::Parameter * const par = static_cast<const decision_tree::classification::Parameter *>(parameter);
+    decision_tree::Pruning treePruning;
+    {
+        auto par1 = dynamic_cast<const decision_tree::classification::interface1::Parameter *>(parameter);
```
[the remark this answers is not in the document]
Comment (orazve) [reply algorithms/kernel/decision_tree/decision_tree_classification_training_input.cpp]: I couldn't find them

File: service/kernel/service_utils.h [hunk 35]
```diff
@@ -139,8 +139,14 @@ ForwardIterator upperBound(ForwardIterator first, ForwardIterator last, const T
 template <CpuType cpu, typename ForwardIterator, typename T, typename Compare>
 ForwardIterator upperBound(ForwardIterator first, ForwardIterator last, const T & value, Compare compare)
 {
-    ForwardIterator it;
     auto count = last - first; // distance.
+    if(count > 0)
+    {
+        if(compare(value, *first)) return first;
+        auto second = first + 1;
```
[the remark this answers is not in the document]
Comment (orazve) [reply service/kernel/service_utils.h]:
```
        auto second = first;
        second++;
```
Like that?

File: algorithms/kernel/classifier/classifier_train.cpp [hunk 36]
```diff
@@ -43,7 +43,17 @@ services::Status Parameter::check() const
...
+services::Status Parameter::check() const
+{
+    DAAL_CHECK_EX(nClasses > 0, services::ErrorIncorrectParameter, services::ParameterName, nClassesStr());
+    DAAL_CHECK_EX(resultsToEvaluate != 0, services::ErrorIncorrectParameter, services::ParameterName, resultsToEvaluateStr())
+    return services::Status();
+}
+}
 namespace training
 {
 namespace interface1
```
Comment (ShvetsKS) [inline algorithms/kernel/classifier/classifier_train.cpp]: On row 106 in current file we always cast to interface2::Parammeter class. That`s wrong.
Branching should be added.

File: service/kernel/service_utils.h (hunk 35 above)
[the remark this answers is not in the document]
Comment (averbukh) [reply service/kernel/service_utils.h]: auto second = first;
++second;

We don't want to create temporary object, right?

File: include/algorithms/logistic_regression/logistic_regression_predict_types.h (hunk 15 above)
Comment (orazve) [reply include/algorithms/logistic_regression/logistic_regression_predict_types.h]: fixed by adding usings from classifier class

File: algorithms/kernel/multiclassclassifier/multiclassclassifier_predict_kernel.h [hunk 37]
```diff
@@ -45,7 +45,7 @@ namespace prediction
 {
 namespace internal
 {
-template<prediction::Method pmethod, training::Method tmethod, typename algorithmFPType, CpuType cpu>
+template<prediction::Method pmethod, training::Method tmethod, typename algorithmFPType, typename ClsType, typename MccParam, CpuType cpu>
```
[the remark this answers is not in the document]
Comment (orazve) [reply algorithms/kernel/multiclassclassifier/multiclassclassifier_predict_kernel.h]: Somewhere in kernel if multiclass classifier mcc abbreviation is already used. So I used it too. Due to it is only template parameter name, I think this point is not so critical. But if don't think so, we can find a better way to name this template parameter.
